src/server.toit

```diff
@@ -175,18 +204,24 @@ class Server:
             else:
               close-logger.debug "connection ended"
           finally:
-            if semaphore_: semaphore_.up  // Up the semaphore when the task ends.
+            task-count_--
+            signal_.raise
         // End of code that can be run in the current task or in a child task.
 
-        parent-task-semaphore = null  // We got this far, the semaphore is ours.
-        if semaphore_:
+        if max-tasks_ > 1:
           task --background handle-connection-closure
         else:
           // For the single-task case, just run the connection in the current task.
           handle-connection-closure.call
+        // At this point the `handle-connection-closure` function is responsible
+        // for releasing the reserved task.
+        need-to-release-reserved-task = false
       finally:
-        // Up the semaphore if we threw before starting the task.
-        if parent-task-semaphore: parent-task-semaphore.up
+        // Release the reserved task if the code threw before we entered
```

Comment:
Don't we release if the code threw as part of calling `handle-connection-closure`? Maybe entered should be exited?